drivers/DeviceKey.cpp

```diff
@@ -0,0 +1,234 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "drivers/DeviceKey.h"
+#include "mbedtls/config.h"
+#include "mbedtls/cmac.h"
+#include <nvstore.h>
+#include "trng_api.h"
+
+namespace mbed {
+
+DeviceKey::DeviceKey()
+{
+    return;
+}
+
+DeviceKey::~DeviceKey()
+{
+    return;
+}
+
+#if !defined(MBEDTLS_CMAC_C)
+#error [NOT_SUPPORTED] MBEDTLS_CMAC_C needs to be enabled for this driver
+#else
+
+int DeviceKey::device_key_derived_key(const unsigned char *salt, size_t isalt_size, unsigned char *output, uint16_t ikey_type)
+{
+    uint32_t key_buff[DEVICE_KEY_32BYTE / sizeof(uint32_t)];
+    size_t actual_size = DEVICE_KEY_32BYTE;
+
+    if (DEVICE_KEY_16BYTE != ikey_type && DEVICE_KEY_32BYTE != ikey_type)
+    {
+        return DEVICEKEY_INVALID_KEY_TYPE;
+    }
+
+    //First try to read the key form NVStore
+    int ret = read_key_from_nvstore(key_buff, actual_size);
+    if (DEVICEKEY_SUCCESS != ret && DEVICEKEY_NOT_FOUND != ret)
+    {
+        return ret;
+    }
+
+    //If the key was not found in NVStore we will create it by using TRNG and then save it to NVStore
+    if (DEVICEKEY_NOT_FOUND == ret) {
+#if defined(DEVICE_TRNG)
+        ret = generate_key_by_trng(key_buff,actual_size);
+        if (DEVICEKEY_SUCCESS != ret) {
+            return ret;
+        }
+
+        ret = device_inject_root_of_trust(key_buff,actual_size);
+        if (DEVICEKEY_SUCCESS != ret) {
+            return ret;
+        }
+#else
+        return DEVICEKEY_NO_KEY_INJECTED;
+#endif
+
+    }
+
+    ret = get_derive_key(key_buff, actual_size, salt, isalt_size, output, ikey_type);
+    return ret;
+}
+
+int DeviceKey::device_inject_root_of_trust(uint32_t *value, size_t isize)
+{
+    size_t key_type;
+    if (isize == 16) {
+        key_type = DEVICE_KEY_16BYTE;
+    } else if (isize == 32) {
+        key_type = DEVICE_KEY_32BYTE;
+    } else {
+        return (DEVICEKEY_INVALID_KEY_SIZE);
+    }
+
+    return write_key_to_nvstore(value, key_type);
+}
+
+int DeviceKey::write_key_to_nvstore(uint32_t *value, size_t isize)
+{
+    if (isize > 32 || isize < 16) {
+        return (DEVICEKEY_INVALID_KEY_SIZE);
+    }
+
+    NVStore &nvstore = NVStore::get_instance();
+    int nvStatus = nvstore.set_once(NVSTORE_KEY_ROT, (uint16_t)isize, value);
+    if (NVSTORE_ALREADY_EXISTS == nvStatus) {
+        return DEVICEKEY_ALREADY_EXIST;
+    }
+
+    if (NVSTORE_WRITE_ERROR == nvStatus || NVSTORE_BUFF_TOO_SMALL == nvStatus) {
+        return DEVICEKEY_SAVE_FAILED;
+    }
+
+    if (NVSTORE_SUCCESS != nvStatus) {
+        return DEVICEKEY_NVSTORE_UNPREDICTABLE_ERROR;
+    }
+
+    return DEVICEKEY_SUCCESS;
+}
+
+int DeviceKey::read_key_from_nvstore(uint32_t *output, size_t &size)
+{
+    uint16_t short_size = size;
+    NVStore &nvstore = NVStore::get_instance();
+    int nvStatus = nvstore.get(NVSTORE_KEY_ROT, short_size, output, short_size);
```

Comment:
Using the same argument for buffer size and returned size parameters is a very bad idea. Please define an additional variable for it (also applies to all other get instances).